### ISSUE TITLE
Using Promise.all inside transactions errors with "Cannot pin multiple connections to the same session"

### ISSUE THREAD

**Issue by fiws** [Automattic/mongoose#14603]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

8.4.0

### Node.js version

22.0.0

### MongoDB server version

7.3.2

### Description

Using running mongo queries/commands in `Promise.all` wrapped in a transaction throws the following error:

```
TypeError: Cannot pin multiple connections to the same session
```

I tested this using an atlas serverless database. Is it not allowed to run queries in parallel when using a transaction?

### Steps to Reproduce

```
import mongoose from 'mongoose';
mongoose.connect(process.env.MONGO_URI || 'mongodb://localhost/transaction-test');

const userSchema = new mongoose.Schema({
  name: String,
});

const User = mongoose.model('User', userSchema);

test('user transaction test', async () => {
  await Instance.db.transaction(async session => {
    return Promise.all([
      User.findOne({}, null, { session }),
      User.findOne({}, null, { session }),
    ]);
  });
});
```

### Expected Behavior

No error.

**Comment by fiws** [comment]
I don't know if this is related but in a more complex application we did also get worrisome side-effects from these kind of transactions, where a `findById` or `findOne` looking up a document that has been deleted in the previous transactions suddenly returns something like this:

```
{
  ok: 1,
  '$clusterTime': {
    clusterTime: new Timestamp({ t: 1715964070, i: 18 }),
    signature: {
      hash: Binary.createFromBase64('QvEPV74zpqEMIRD5lLDYPkfxsV8=', 0),
      keyId: new Long('7330005257113567234')
    }
  },
  operationTime: new Timestamp({ t: 1715964070, i: 18 })
}
```

This looks like some mongo internals that should never be returned by `findById`.

Might be another bug tho, and I'm still trying to get this into a small reproducible snippet.

**Comment by IslandRhythms** [comment]
I noticed in your repo script you're using a local connection. Transactions can only be used on replica sets. Just want to confirm that you tested on a replica set before creating the issue.

**Comment by fiws** [comment]
Yep. I used a serverless MongoDB atlas for testing.

**Comment by vkarpov15** [comment]
I opened issue in MongoDB Jira here: https://jira.mongodb.org/browse/NODE-6190

**Comment by vkarpov15** [comment]
As a temporary workaround, you can use the following:

```
  await mongoose.connection.transaction(async session => {
    const docs = await Promise.all([
      User.findOne({}, null, { session }),
      sleep(0).then(() => User.findOne({}, null, { session })),
    ]);
    console.log(docs);
  });

async function sleep(ms) {
  await new Promise(resolve => setTimeout(resolve, ms));
}
```

**Comment by vkarpov15** [comment]
It looks like this is expected behavior. From the [JIRA ticket](https://jira.mongodb.org/browse/NODE-6190):

![image](https://github.com/Automattic/mongoose/assets/1620265/d1ab1d90-2563-4905-b4f7-8ef559e5fa51)

[Here's the docs link](https://www.mongodb.com/docs/drivers/node/current/fundamentals/transactions/#convenient-transaction-api). I'm unfortunately going to have to close this issue: Mongoose can't support parallel requests with the same transaction if the MongoDB server doesn't support it.

**Comment by fiws** [comment]
Thanks for researching this and creating the upstream issue.

Maybe some documentation on mongoose for this would be nice. The big problem is that this is undefined behaviour. It does not throw the error every time. We have `Promise.all` calls that seemingly work (in production). We noticed [super weird behaviour](https://github.com/Automattic/mongoose/issues/14603#issuecomment-2117998112) recently though, which is the whole reason why i created this issue.

**Comment by maxeth** [comment]
With `.withTransaction()` this issue only appears when the first operation of the transaction is part of the `Promise.all()`.

So this works

```ts
      const a = await ModelA.find({}, {}, { session });
      const [b,c] = await Promise.all([...])
```

while it doesn't if a is included in the Promise.all() array when using a Serverless instance specifically.

 noticed the **client becoming fully unresponsive** at random times after executing a few transactions. Calling an db operation like `MyModel.find()` just resulted in indefinitely pending promises until I restarted the server.

This only happens with transactions/sessions and with an Atlas Serverless instance so I think it may somehow be related to this thread.

---

A bizarre solution to this is ending the session manually, like

```ts
await session.withTransaction(async () => {}).then(() => session.endSession())
```

even though the session with `.transaction()` and `.withTransaction()` is already being ended [automatically](https://github.com/Automattic/mongoose/blob/master/lib/connection.js#L557) /  #13719  by mongoose!? 🤔

I ran the same transaction thousands of times with this and it works, while it starts hanging after the ~10th attempt when not ending the session manually.

Seems to me that Atlas Serverless instances aren't reliably usable with mongoose.
